File: sdks/go/protocol.go
package mdp

import (
	"encoding/json"
	"errors"
	"fmt"
)

type ClientToServerMessage map[string]any

type ServerToClientMessage interface {
	messageType() string
}

type PingMessage struct {
	Timestamp int64
}

func (message PingMessage) messageType() string { return "ping" }

type PongMessage struct {
	Timestamp int64
}

func (message PongMessage) messageType() string { return "pong" }

type CallClientMessage struct {
	RequestID string
	ClientID  string
	Method    HttpMethod
	Path      string
	Params    map[string]any
	Query     map[string]any
	Body      any
	Headers   map[string]string
	Auth      *AuthContext
}

func (message CallClientMessage) messageType() string { return "callClient" }

func BuildRegisterClientMessage(client ClientDescriptor, auth *AuthContext) ClientToServerMessage {
	payload := ClientToServerMessage{
		"type":   "registerClient",
		"client": client.ToMap(),
	}
	if auth != nil {
		payload["auth"] = auth
	}
	return payload
}

func BuildUpdateClientCatalogMessage(clientID string, paths []PathDescriptor) ClientToServerMessage {
	serializedPaths := make([]map[string]any, 0, len(paths))
	for _, path := range paths {
		serializedPaths = append(serializedPaths, path.ToMap())
	}
	return ClientToServerMessage{
		"type":     "updateClientCatalog",
		"clientId": clientID,
		"paths":    serializedPaths,
	}
}

func BuildUnregisterClientMessage(clientID string) ClientToServerMessage {
	return ClientToServerMessage{
		"type":     "unregisterClient",
		"clientId": clientID,
	}
}

func BuildCallClientResultMessage(requestID string, ok bool, data any, protocolError *SerializedError) ClientToServerMessage {
	payload := ClientToServerMessage{
		"type":      "callClientResult",
		"requestId": requestID,
		"ok":        ok,
	}
	if ok {
		payload["data"] = data
	} else if protocolError != nil {
		payload["error"] = protocolError
	}
	return payload
}

func BuildPongMessage(timestamp int64) ClientToServerMessage {
	return ClientToServerMessage{
		"type":      "pong",
		"timestamp": timestamp,
	}
}

func ParseServerToClientMessage(raw []byte) (ServerToClientMessage, error) {
	var envelope map[string]any
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil, err
	}

	messageType, ok := envelope["type"].(string)
	if !ok {
		return nil, errors.New("missing message type")
	}

	switch messageType {
	case "ping":
		timestamp, ok := asInt64(envelope["timestamp"])
		if !ok {
			return nil, errors.New("invalid ping payload")
		}
		return PingMessage{Timestamp: timestamp}, nil
	case "pong":
		timestamp, ok := asInt64(envelope["timestamp"])
		if !ok {
			return nil, errors.New("invalid pong payload")
		}
		return PongMessage{Timestamp: timestamp}, nil
	case "callClient":
		requestID, ok := envelope["requestId"].(string)
		if !ok {
			return nil, errors.New("missing requestId")
		}
		clientID, ok := envelope["clientId"].(string)
		if !ok {
			return nil, errors.New("missing clientId")
		}
		method, ok := envelope["method"].(string)
		if !ok {
			return nil, errors.New("missing method")
		}
		path, ok := envelope["path"].(string)
		if !ok || !IsConcretePath(path) {
			return nil, errors.New("invalid path")
		}
		return CallClientMessage{
			RequestID: requestID,
			ClientID:  clientID,
			Method:    HttpMethod(method),
			Path:      path,
			Params:    asMap(envelope["params"]),
			Query:     asMap(envelope["query"]),
			Body:      envelope["body"],
			Headers:   asStringMap(envelope["headers"]),
			Auth:      asAuthContext(envelope["auth"]),
		}, nil
	default:
		return nil, fmt.Errorf("unsupported message type %q", messageType)
	}
}

func asMap(value any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	if typed, ok := value.(map[string]any); ok {
		return typed
	}
	return map[string]any{}
}

func asStringMap(value any) map[string]string {
	if value == nil {
		return map[string]string{}
	}
	typed, ok := value.(map[string]any)
	if !ok {
		return map[string]string{}
	}
	result := make(map[string]string, len(typed))
	for key, entry := range typed {
		if stringValue, ok := entry.(string); ok {
			result[key] = stringValue
		}
	}
	return result
}

func asInt64(value any) (int64, bool) {
	switch typed := value.(type) {
	case float64:
		return int64(typed), true
	case int64:
		return typed, true
	case int:
		return int64(typed), true
	default:
		return 0, false
	}
}

func asAuthContext(value any) *AuthContext {
	typed, ok := value.(map[string]any)
	if !ok {
		return nil
	}
	auth := &AuthContext{}
	if scheme, ok := typed["scheme"].(string); ok {
		auth.Scheme = scheme
	}
	if token, ok := typed["token"].(string); ok {
		auth.Token = token
	}
	auth.Headers = asStringMap(typed["headers"])
	auth.Metadata = asMap(typed["metadata"])
	return auth
}

File: sdks/rust/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::error::MdpClientError;
use crate::models::{AuthContext, ClientDescriptor, HttpMethod, SerializedError};
use crate::path_utils::is_concrete_path;

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(tag = "type")]
pub enum ClientToServerMessage {
    #[serde(rename = "registerClient")]
    RegisterClient {
        client: ClientDescriptor,
        #[serde(skip_serializing_if = "Option::is_none")]
        auth: Option<AuthContext>,
    },
    #[serde(rename = "updateClientCatalog")]
    UpdateClientCatalog {
        #[serde(rename = "clientId")]
        client_id: String,
        paths: Vec<crate::models::PathDescriptor>,
    },
    #[serde(rename = "unregisterClient")]
    UnregisterClient {
        #[serde(rename = "clientId")]
        client_id: String,
    },
    #[serde(rename = "callClientResult")]
    CallClientResult {
        #[serde(rename = "requestId")]
        request_id: String,
        ok: bool,
        #[serde(skip_serializing_if = "Option::is_none")]
        data: Option<Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<SerializedError>,
    },
    #[serde(rename = "pong")]
    Pong { timestamp: i64 },
}

#[derive(Clone, Debug, PartialEq)]
pub enum ServerToClientMessage {
    CallClient(CallClientRequest),
    Ping { timestamp: i64 },
    Pong { timestamp: i64 },
}

#[derive(Clone, Debug, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallClientRequest {
    pub request_id: String,
    pub client_id: String,
    pub method: HttpMethod,
    pub path: String,
    pub params: Option<Map<String, Value>>,
    pub query: Option<Map<String, Value>>,
    pub body: Option<Value>,
    pub headers: Option<std::collections::HashMap<String, String>>,
    pub auth: Option<AuthContext>,
}

impl ServerToClientMessage {
    pub fn from_text(raw: &str) -> Result<Self, MdpClientError> {
        let value: Value = serde_json::from_str(raw)?;
        Self::from_value(value)
    }

    pub fn from_value(value: Value) -> Result<Self, MdpClientError> {
        let object = value
            .as_object()
            .ok_or_else(|| MdpClientError::Protocol("invalid MDP message payload".to_string()))?;

        let message_type = object
            .get("type")
            .and_then(Value::as_str)
            .ok_or_else(|| MdpClientError::Protocol("missing message type".to_string()))?;

        match message_type {
            "ping" => Ok(Self::Ping {
                timestamp: object
                    .get("timestamp")
                    .and_then(Value::as_i64)
                    .ok_or_else(|| MdpClientError::Protocol("invalid ping payload".to_string()))?,
            }),
            "pong" => Ok(Self::Pong {
                timestamp: object
                    .get("timestamp")
                    .and_then(Value::as_i64)
                    .ok_or_else(|| MdpClientError::Protocol("invalid pong payload".to_string()))?,
            }),
            "callClient" => {
                let request: CallClientRequest = serde_json::from_value(value)?;
                if !is_concrete_path(&request.path) {
                    return Err(MdpClientError::Protocol("invalid callClient path".to_string()));
                }
                Ok(Self::CallClient(request))
            }
            other => Err(MdpClientError::Protocol(format!(
                "unsupported server message type `{other}`"
            ))),
        }
    }
}
